File: pa2/List.c
/*
 * Name: Mohammad Mirabian
 * CruzID: mmirabia
 * File Name: List.c
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdbool.h>
#include "List.h"

typedef struct NodeObj{
     int data;
     struct NodeObj *prev;
     struct NodeObj *next;
} NodeObj;

typedef struct NodeObj* Node;

typedef struct ListObj{
     Node front;
     Node curr;
     Node back;
     int length;
} ListObj;

//node constructor
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   return(N);
}

//node destructor
void freeNode(Node* pN){
    if( pN!=NULL && *pN!=NULL ){
          free(*pN);
          *pN = NULL;
     }
}

//List constructor
List newList(void){
     List L = malloc(sizeof(ListObj));
     L->front = L->curr = L->back = NULL;
     L->length = 0;
     return L;
}

//List destructor
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) {
        while(length(*pL) != 0) {
           deleteFront(*pL);
      }
      free(*pL);
      *pL = NULL;
   }
}
// Access functions -----------------------------------------------------------


int length(List L){
     if(L->front == NULL) return 0;
     return L->length;
}


int index(List L){
     Node a = L->front;
     int count = 0;
     if(L->curr != NULL){
       if(L->curr->data == a->data){ return (count + 1); }
       while(a != NULL){
          count++;
          if(L->curr->data == a->next->data){
             return count;
          }
          a = a->next;
          }
     }
     return -1;
}


int front(List L){
    if(length(L) > 0) return L->front->data;
    else{
       return -1;}
}


int back(List L){
    if(length(L) > 0){ return L->back->data;
    }else{
       return -1;
    }
}


int get(List L){
    if((index(L) > -1) && (length(L) > 0)) return L->curr->data;
    else{
       fprintf(stderr,"cursor is off the List");
       return -1;
    }
}

int equals(List A, List B){
      bool equalsflag = true;
      Node a = A->front;
      Node b = B->front;
      while(equalsflag && a != NULL){
            equalsflag = (a->data == b->data);
            a = a->next;
            b = b->next;
          }
      return equalsflag;
}

// Manipulation procedures ----------------------------------------------------

void clear(List L){
    if(L->front == NULL) return;
    else{
        Node fronttmp = L->front;
        while(fronttmp != NULL){
           Node tmp = fronttmp->next;
           free(fronttmp);
           fronttmp = tmp;
    }
    L->front = L->back = L->curr = NULL;
    L->length = 0;
    }
}

void moveFront(List L){
    if(length(L) > 0){
        L->curr = L->front;
    }
}
void moveBack(List L){
    if(length(L) > 0){
        L->curr = L->back;
    }
}

void movePrev(List L){
    Node a;
    if((length(L) == 0)||(L->curr == NULL)){
        //opposite if for the preconditions^
        fprintf(stderr,"Empty list movePrev()");
        exit(EXIT_FAILURE);
    }else{
      a = L->curr;
      L->curr = a->prev;
    }
}


void moveNext(List L){
     Node a;
     if(length(L) == 0){
        fprintf(stderr,"empty List moveNext()");
        exit(EXIT_FAILURE);
     }else{
        a = L->curr;
        L->curr = a->next;
     }
}


void prepend(List L, int data){
     if(L == NULL){
        fprintf(stderr,"error L == NULL prepend");
        return;}
     Node n = malloc(sizeof(struct NodeObj));
     n->data = data;
     n->prev = NULL;
     n->next = L->front;
     if(L->front == NULL){
        L->back = n;
     }else{
        L->front->prev = n;
     }
     L->front = n;
     L->length++;
}


void append(List L, int data){
    if(L == NULL){
        fprintf(stderr,"error L == NULL prepend");
        return;}
     Node n = malloc(sizeof(struct NodeObj));
     n->data = data;
     n->next = NULL;
     n->prev = L->back;
     if(L->back == NULL){
        L->front = n;
    }else{
        L->back->next = n;
    }
    L->back = n;
    L->length++;
}


void insertBefore(List L, int data){
     if(length(L) < 0 && index(L) <= 0){
          fprintf(stderr, "empty List insertBefore()");
          exit(EXIT_FAILURE);
     }
     L->length++;
     Node before = newNode(data);
     before->next = L->curr;
     if(L->curr != L->front){
          L->curr->prev->next = before;
          before->prev = L->curr->next;
     }
     L->curr->prev = before;
     if(L->curr == L->back) L->front = before;
}


void insertAfter(List L, int data){
     if(length(L) == 0){
          fprintf(stderr, "empty List InsertAfter()");
          exit(EXIT_FAILURE);}
     L->length++;
     Node after = newNode(data);
     after->prev = L->curr;
     if(L->curr != L->back){
           L->curr->next->prev = after;
           after->next = L->curr->next;
     }
     L->curr->next = after;
     if(L->curr == L->back) L->back = after;
}


void deleteFront(List L){
     Node tmp;
     if(length(L) == 0){
            fprintf(stderr,"empty list deleteFront()");
            exit(EXIT_FAILURE);}
     if(length(L) > 1){
        tmp = L->front;
        L->front = L->front->next;
        L->front->prev = NULL;}
     else{
        tmp = L->front;
        L->front = L->back = NULL;}
     freeNode(&tmp);
     L->length--;
}


void deleteBack(List L){
     Node tmp;
     if(length(L) == 0) {
             fprintf(stderr,"empty list deleteBack()");
             exit(EXIT_FAILURE);}
     if(length(L) > 1){
        tmp = L->back;
        L->back = L->back->prev;
        L->back->next = NULL;}
     else{
        tmp = L->back;
        L->back = L->front = NULL;}
     freeNode(&tmp);
     L->length--;
}


void delete(List L){
     Node tmp;
     if(length(L) <= 0 && index(L) < 0)
        exit(EXIT_FAILURE);
     else{
        tmp = L->curr->prev;
        tmp->next = L->curr->next;
        if(tmp != NULL){
           L->curr = L->curr->next;
           L->curr->prev = tmp;
        }
     }
}

// Other operations -----------------------------------------------------------

void printList(FILE* out, List L){
    if(L == NULL){
       fprintf(stderr,"Empty List printlist()");
       exit(EXIT_FAILURE);
    }
    if(out == NULL){
       fprintf(stderr,"no out FILE printlist()");
       return;
    }
    Node n = L->front;
    while(n != NULL){
         fprintf(out, "%d ",n->data);
         n = n->next;
    }
}


List copyList(List L){
    List cpy = newList();
    Node A = L->front;
    while(A != NULL){
        append(cpy, A->data);
        A = A->next;
    }
    return cpy;
}

File: pa3/List.java
/*
 * Name: Mohammad Mirabian
 * CruzID: mmirabia
 * File Name: List.java
 * File Purpose: ADT
 *7/2/2015
 */
public class List {

	// private inner Node class that defines a Node inside the list
	private class Node {
		Object data;
		Node next;
		Node previous;

		// constructor for Node class
		Node(Object givenData) {
			data = givenData;
			next = null;
			previous = null;
		}

		Node(Object givenData, Node pre, Node nxt) {
			data = givenData;
			previous = pre;
			next = nxt;
		}

		public String toString() {
			return "" + data;
		}
	}

	// class List fields
	private Node front;
	private Node back;
	private Node cursor;
	private int length;

	// private int index;

	// class List Constructor
	List() {
		front = null;
		back = null;
		cursor = null;
		length = 0;
		// index = -1;
	}

	// Access functions
	int length() { // Returns the number of elements in this List.
		return length;

	}

	int index() { // If cursor is defined, returns the index of the cursor
					// element,
		// otherwise returns -1.
		int index = -1;
		if (cursor == null)
			return index;
		for (Node tmp = front; tmp != null; tmp = tmp.next) {
			index++;
			if (tmp == cursor)
				break;
		}
		return index;
	}

	Object front() { // Returns front element. Pre: length()>0
		if (length() > 0) {
			return front.data;
		}
		return -1;
	}

	Object back() { // Returns back element. Pre: length()>0
		if (length() > 0) {
			return back.data;
		}
		return -1;
	}

	Object get() { // Returns cursor element. Pre: length()>0, index()>=0
		if (length() > 0 && index() >= 0)
			return cursor.data;
		return -1;
	}

	public boolean equals(Object L) { // Returns true if this List and L are the
										// same
		// integer
		// sequence. The cursor is ignored in both lists.
		Node myNode = front;
		Node givenNode = ((List) L).front;

		while (myNode != null && givenNode != null) {
			if (myNode.data != givenNode.data) {
				return false;
			}
			myNode = myNode.next;
			givenNode = givenNode.next;
		}
		if (myNode != null || givenNode != null) {
			return false;
		}

		else {
			return true;
		}
	}

	// Manipulation procedures
	void clear() { // Resets this List to its original empty state.
		cursor = null;
		front = null;
		back = null;
		length = 0;

	}

	void moveFront() { // If List is non-empty, places the cursor under the
						// front element,
		// otherwise does nothing.
		if (length > 0) {
			cursor = front;
		}
	}

	void moveBack() { // If List is non-empty, places the cursor under the back
						// element,
		// otherwise does nothing.
		if (length > 0) {
			cursor = back;
		}
	}

	void movePrev() { // If cursor is defined and not at front, moves cursor one
						// step toward
		// front of this List, if cursor is defined and at front, cursor becomes
		// undefined, if cursor is undefined does nothing.
		int index = index();
		index = index();
		if (cursor != null) {
			if (cursor != front) {
				cursor = cursor.previous;
			} else if (cursor == front) {
				cursor = null;
			}
		}
	}

	void moveNext() { // If cursor is defined and not at back, moves cursor one
						// step toward
		// back of this List, if cursor is defined and at back, cursor becomes
		// undefined, if cursor is undefined does nothing.
		int index = index();
		if (cursor != null) {
			if (cursor != back) {
				cursor = cursor.next;
			} else if (cursor == back) {
				cursor = null;
			}
		}
	}

	void prepend(Object data) { // Insert new element into this List. If List is
								// non-empty,
		// insertion takes place before front element.
		length++;
		Node N = new Node(data, null, front);
		if (front != null) {
			front.previous = N;
		} else {
			back = N;
		}
		front = N;
	}

	void append(Object data) { // Insert new element into this List. If List is
								// non-empty,
		// insertion takes place after back element.
		length++;
		Node N = new Node(data, back, null);
		if (back != null) {
			back.next = N;
		} else {
			front = N;
		}
		back = N;
	}

	void insertBefore(Object data) { // Insert new element before cursor.
		// Pre: length()>0, index()>=0
		length++;
		if (index() >= 0 && length() >= 0) {
			Node N = new Node(data, cursor.previous, cursor);
			if (cursor.previous != null) {
				cursor.previous.next = N;
			} else {
				front = N;
			}

			cursor.previous = N;
		}
	}

	void insertAfter(Object data) { // Inserts new element after cursor.
		// Pre: length()>0, index()>=0
		length++;
		if (index() >= 0 && length() >= 0) {
			Node N = new Node(data, cursor, cursor.next);
			if (cursor.next != null) {
				cursor.next.previous = N;
			} else {
				back = N;
			}

			cursor.next = N;
		}
	}

	void deleteFront() { // Deletes the front element. Pre: length()>0
		length--;
		if (length() > 0) {
			if (cursor == front)
				cursor = null;
			front = front.next;
			front.previous = null;
		}

	}

	void deleteBack() { // Deletes the back element. Pre: length()>0
		length--;
		if (length() > 0) {
			if (cursor == back)
				cursor = null;
			back = back.previous;
			back.next = null;
		}

	}

	void delete() // Deletes cursor element, making
	{
		if (cursor == front) {
			deleteFront();
		} else if (cursor == back) {
			deleteBack();
		}

		if (length() > 0 && index() > 0) {
			length--;
			cursor.next.previous = cursor.previous;
			cursor.previous.next = cursor.next;
			cursor = null;
		}

	}

	// Other methods
	public String toString() { // Overrides Object's toString method. Returns a
								// String
		// representation of this List consisting of a space
		// separated sequence of integers, with front on left.
		String str = "";
		Node temp = front;
		while (temp != null) {
			str += temp + " ";
			temp = temp.next;
		}
		return str;
	}

}
